### PR TITLE
Return when current branch token changed in poll mutable state API 

### PR DESCRIPTION
**Pull request by yux0**
Return when current branch token changed in poll mutable state API for 3+DC feature

### CODE REVIEW

File: service/frontend/workflowHandler.go [hunk 1]
```diff
@@ -1754,6 +1767,12 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
 		token.PersistenceToken = nil
 	}
 
+	// calculate event store version based on if branch token exist
+	token.EventStoreVersion = persistence.EventStoreVersionV2
```
Comment (wxing1292) [inline service/frontend/workflowHandler.go]: can you move this logic into `queryHistory` ?

File: service/frontend/workflowHandler.go [hunk 2]
```diff
@@ -2882,13 +2899,12 @@ func (wh *WorkflowHandler) getHistory(
...
 	nextPageToken []byte,
 	transientDecision *gen.TransientDecisionInfo,
-	eventStoreVersion int32,
 	branchToken []byte,
 ) (*gen.History, []byte, error) {
 
 	historyEvents := []*gen.HistoryEvent{}
 	var size int
-	if eventStoreVersion == persistence.EventStoreVersionV2 {
+	if branchToken != nil && len(branchToken) != 0 {
```
Comment (wxing1292) [inline service/frontend/workflowHandler.go]: `len(branchToken)` should be enough

File: service/history/conflictResolver_test.go [hunk 3]
```diff
@@ -361,7 +361,7 @@ func (s *conflictResolverSuite) TestReset() {
 			ExecutionInfo:  &persistence.WorkflowExecutionInfo{},
 			ExecutionStats: &persistence.ExecutionStats{},
 		},
-	}, nil).Once() // return empty resoonse since we are not testing the load
+	}, nil).Maybe() // return empty resoonse since we are not testing the load
```
Comment (wxing1292) [inline service/history/conflictResolver_test.go]: why changing this line?

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -567,7 +569,13 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
 	if err != nil {
 		return nil, err
 	}
-
```
Comment (wxing1292) [inline service/history/historyEngine.go]: how about
```
if request.CurrentBranchToken == nil {
  request.CurrentBranchToken = response.CurrentBranchToken
}

if !bytes.Equal(request.CurrentBranchToken, response.CurrentBranchToken) {
  return nil, &workflow.CurrentBranchChangedError{
  Message:  "current branch token and request branch token doesn't match.",
  CurrentBranchToken: response.CurrentBranchToken}
}
```

File: service/history/historyEngine.go [hunk 5]
```diff
@@ -590,8 +598,14 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
 		if err != nil {
 			return nil, err
 		}
-
-		if expectedNextEventID < response.GetNextEventId() ||
+		// check again if the current branch token changed
+		if request.CurrentBranchToken != nil && !bytes.Equal(request.CurrentBranchToken, currBranchToken) {
```
Comment (wxing1292) [inline service/history/historyEngine.go]:
```
if !bytes.Equal(request.CurrentBranchToken, response.CurrentBranchToken) {
  return nil, &workflow.CurrentBranchChangedError{
  Message:  "current branch token and request branch token doesn't match.",
  CurrentBranchToken: response.CurrentBranchToken}
}
```

File: service/history/historyEngine.go [hunk 6]
```diff
@@ -590,8 +598,14 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
...
+			return nil, &workflow.CurrentBranchChangedError{
+				Message:            "current branch token and request branch token doesn't match.",
+				CurrentBranchToken: currBranchToken}
+		}
+		if !bytes.Equal(response.CurrentBranchToken, currBranchToken) ||
```
Comment (wxing1292) [inline service/history/historyEngine.go]:
```
if expectedNextEventID < response.GetNextEventId() || !response.GetIsWorkflowRunning() {
  return response, nil
}
```

File: service/history/historyEngine.go [hunk 7]
```diff
@@ -609,7 +623,15 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
 				response.NextEventId = common.Int64Ptr(event.nextEventID)
 				response.IsWorkflowRunning = common.BoolPtr(event.isWorkflowRunning)
 				response.PreviousStartedEventId = common.Int64Ptr(event.previousStartedEventID)
-				if expectedNextEventID < response.GetNextEventId() ||
+				response.WorkflowState = common.Int32Ptr(int32(event.workflowExecutionState))
+				response.WorkflowCloseState = common.Int32Ptr(int32(event.workflowExecutionCloseState))
+				if request.CurrentBranchToken != nil && !bytes.Equal(request.CurrentBranchToken, event.currentBranchToken) {
```
Comment (wxing1292) [inline service/history/historyEngine.go]: same as above

File: service/history/historyEngine.go [hunk 8]
```diff
@@ -609,7 +623,15 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
...
+					return nil, &workflow.CurrentBranchChangedError{
+						Message:            "Current branch token and request branch token doesn't match.",
+						CurrentBranchToken: currBranchToken}
+				}
+				if !bytes.Equal(event.currentBranchToken, currBranchToken) ||
```
Comment (wxing1292) [inline service/history/historyEngine.go]: same as above

File: service/history/historyEngineInterfaces.go [hunk 9]
```diff
@@ -33,12 +33,15 @@ import (
...
-		lastFirstEventID       int64
-		nextEventID            int64
-		previousStartedEventID int64
-		isWorkflowRunning      bool
-		timestamp              time.Time
+		id                          definition.WorkflowIdentifier
+		lastFirstEventID            int64
+		nextEventID                 int64
+		previousStartedEventID      int64
+		isWorkflowRunning           bool
```
Comment (wxing1292) [inline service/history/historyEngineInterfaces.go]: you can delete this field

File: service/history/historyEventNotifier_test.go [hunk 10]
```diff
@@ -21,6 +21,7 @@
 package history
 
 import (
+	"github.com/uber/cadence/common/persistence"
```
Comment (wxing1292) [inline service/history/historyEventNotifier_test.go]: fmt

File: service/frontend/workflowHandler.go (hunk 1 above)
Comment (yux0) [reply service/frontend/workflowHandler.go]: done

File: service/frontend/workflowHandler.go (hunk 2 above)
Comment (yux0) [reply service/frontend/workflowHandler.go]: done

File: service/history/conflictResolver_test.go (hunk 3 above)
Comment (yux0) [reply service/history/conflictResolver_test.go]: Good catch. I was debug this. Revert it

File: service/history/historyEngine.go (hunk 5 above)
Comment (yux0) [reply service/history/historyEngine.go]: synced offline. updated

File: service/history/historyEngine.go (hunk 6 above)
Comment (yux0) [reply service/history/historyEngine.go]: updated

File: service/history/historyEventNotifier_test.go (hunk 10 above)
Comment (yux0) [reply service/history/historyEventNotifier_test.go]: done

File: service/history/historyEngineInterfaces.go (hunk 9 above)
Comment (yux0) [reply service/history/historyEngineInterfaces.go]: removed

**Pull request merged**